// @jxkang/web-cmpt/lib/select-city/select-city.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _select = require('antd/lib/select');

var _select2 = _interopRequireDefault(_select);

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _slicedToArray = function () { function sliceIterator(arr, i) { var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"]) _i["return"](); } finally { if (_d) throw _e; } } return _arr; } return function (arr, i) { if (Array.isArray(arr)) { return arr; } else if (Symbol.iterator in Object(arr)) { return sliceIterator(arr, i); } else { throw new TypeError("Invalid attempt to destructure non-iterable instance"); } }; }(); /**
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          * @Author: 谭生虎 TanShenghu tanshenghu@163.com
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          * @Update: 2019-12-17
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          * @Description: 行政区划选择组件
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          */


require('antd/lib/select/style');

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _utils = require('@jxkang/utils');

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _style = require('./style');

var _style2 = _interopRequireDefault(_style);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _toConsumableArray(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = Array(arr.length); i < arr.length; i++) { arr2[i] = arr[i]; } return arr2; } else { return Array.from(arr); } }

var noop = function noop() {};
var fetchUrl = function fetchUrl(reqModel) {
  return (0, _utils.$post)('/productservice/sys/getAreaList', reqModel);
};
// 所有省市区的数据
var cityData = null;

var SelectCity = function SelectCity(_ref) {
  var _ref$url = _ref.url,
      url = _ref$url === undefined ? fetchUrl : _ref$url,
      className = _ref.className,
      provinceClassName = _ref.provinceClassName,
      cityClassName = _ref.cityClassName,
      countyClassName = _ref.countyClassName,
      leafKeyName = _ref.leafKeyName,
      _ref$onChange = _ref.onChange,
      onChange = _ref$onChange === undefined ? noop : _ref$onChange,
      value = _ref.value,
      _ref$width = _ref.width,
      width = _ref$width === undefined ? [115, 115, 115] : _ref$width,
      _ref$onOk = _ref.onOk,
      onOk = _ref$onOk === undefined ? noop : _ref$onOk,
      _ref$onRef = _ref.onRef,
      onRef = _ref$onRef === undefined ? noop : _ref$onRef,
      _ref$placeholder = _ref.placeholder,
      placeholder = _ref$placeholder === undefined ? ['省份', '地市', '区县'] : _ref$placeholder,
      _ref$disabled = _ref.disabled,
      disabled = _ref$disabled === undefined ? [false, false, false] : _ref$disabled,
      onlyCity = _ref.onlyCity;

  var _useState = (0, _react.useState)([[], [], []]),
      _useState2 = _slicedToArray(_useState, 2),
      datas = _useState2[0],
      setDatas = _useState2[1];

  var _useState3 = (0, _react.useState)([]),
      _useState4 = _slicedToArray(_useState3, 2),
      selected = _useState4[0],
      setSelected = _useState4[1];

  var _useState5 = (0, _react.useState)(value || []),
      _useState6 = _slicedToArray(_useState5, 2),
      selectedId = _useState6[0],
      setSelectedId = _useState6[1];

  var foreign = {
    forceUpdate: function forceUpdate() {
      if (Array.isArray(value) && value.length) {
        setSelectedId(value);
        value.forEach(function (item, index) {
          findCity({ areaCode: item, level: index });
        });
      }
    }
  };

  (0, _react.useEffect)(function () {
    // getData({ parentCode: 0, level: 0 });
    getData({
      parentCode: null
    });
  }, []);

  var getData = function getData(params) {
    var firstFill = function firstFill(data) {
      datas.forEach(function (item) {
        item.length = 0;
      });
      datas[0] = data.filter(function (v) {
        return v.parentCode == '0';
      });
      setDatas([].concat(_toConsumableArray(datas)));

      // 数据回填
      foreign.forceUpdate();
    };

    if (cityData) {
      firstFill(cityData);
    } else {
      var myDB = new _utils.Idb('web-cmpt-tsh', 'city-region');
      myDB.connction().then(function () {
        return myDB.select('city-region', 1);
      }).then(function (resDB) {
        if (resDB) {
          cityData = resDB.data;
          firstFill(cityData);
        } else {
          url(params).then(function (resModel) {
            if (resModel && Array.isArray(resModel.records)) {
              cityData = resModel.records;
              firstFill(cityData);
              // 保存至数据库中
              myDB.insert('city-region', {
                author: '谭生虎 - 福虎',
                desc: '京小康地区行政区划数据结构',
                email: 'tanshenghu@163.com',
                data: resModel.records
              });
            }
          });
        }
      });
    }
  };

  /**
   * 切换城市
   */
  var onChangeCity = function onChangeCity(id, level) {
    // 如果有叶子节点，优先用叶子节点作为判断依据，否则只能通过层级判断
    var lineData = datas[level].find(function (v) {
      return v.areaCode === id;
    });
    onChange(lineData, level, datas);
    selected[level] = lineData;
    selectedId[level] = id;
    // 清空操作
    [0, 1, 2].filter(function (v) {
      return v > level;
    }).forEach(function (item) {
      datas[item] = [];
      selectedId[item] = undefined;
    });
    //
    setSelected(selected);
    setSelectedId([].concat(_toConsumableArray(selectedId)));

    if (leafKeyName) {
      if (lineData[leafKeyName]) {
        onOk(selected, selectedId);
      } else {
        findCity({ areaCode: lineData.areaCode, level: level });
      }
    } else if (level === 2 || onlyCity && level === 1) {
      onOk(selected, selectedId);
    } else {
      findCity({ areaCode: lineData.areaCode, level: level });
    }
  };

  function findCity(_ref2) {
    var areaCode = _ref2.areaCode,
        level = _ref2.level;

    datas[level + 1] = cityData.filter(function (v) {
      return v.parentCode == areaCode;
    });
    setDatas([].concat(_toConsumableArray(datas)));
  }

  onRef(foreign);

  return _react2.default.createElement(
    'section',
    { className: className },
    _react2.default.createElement(
      _select2.default,
      { value: selectedId[0], placeholder: placeholder[0], disabled: disabled[0], style: { width: width[0] }, className: (0, _classnames2.default)(_style2.default.province, provinceClassName), onChange: function onChange(v) {
          return onChangeCity(v, 0);
        } },
      Array.isArray(datas[0]) && datas[0].map(function (item, index) {
        return _react2.default.createElement(
          _select2.default.Option,
          { key: index, value: item.areaCode },
          item.areaName
        );
      })
    ),
    _react2.default.createElement(
      _select2.default,
      { value: selectedId[1], placeholder: placeholder[1], disabled: disabled[1], style: { width: width[1] }, className: (0, _classnames2.default)(_style2.default.city, cityClassName), onChange: function onChange(v) {
          return onChangeCity(v, 1);
        } },
      Array.isArray(datas[1]) && datas[1].map(function (item, index) {
        return _react2.default.createElement(
          _select2.default.Option,
          { key: index, value: item.areaCode },
          item.areaName
        );
      })
    ),
    onlyCity ? null : _react2.default.createElement(
      _select2.default,
      { value: selectedId[2], placeholder: placeholder[2], disabled: disabled[2], style: { width: width[2] }, className: (0, _classnames2.default)(_style2.default.county, countyClassName), onChange: function onChange(v) {
          return onChangeCity(v, 2);
        } },
      Array.isArray(datas[2]) && datas[2].map(function (item, index) {
        return _react2.default.createElement(
          _select2.default.Option,
          { key: index, value: item.areaCode },
          item.areaName
        );
      })
    )
  );
};

/**
 * 通过指定id 查找中文名称
 */
SelectCity.findValue = function (_ref3) {
  var value = _ref3.value,
      _ref3$params = _ref3.params,
      params = _ref3$params === undefined ? {} : _ref3$params,
      _ref3$ofetch = _ref3.ofetch,
      ofetch = _ref3$ofetch === undefined ? fetchUrl : _ref3$ofetch,
      _ref3$callback = _ref3.callback,
      callback = _ref3$callback === undefined ? noop : _ref3$callback;

  if (Array.isArray(value)) {
    var findVal = function findVal(data) {
      var result = value.map(function (item) {
        return data.filter(function (v) {
          return v.areaCode == item;
        })[0];
      });
      callback(result.map(function (v) {
        return v.areaName;
      }), result);
    };

    if (cityData) {
      findVal(cityData);
    } else {
      ofetch(_extends({
        parentCode: null
      }, params)).then(function (resModel) {
        if (resModel && Array.isArray(resModel.records)) {
          cityData = resModel.records;
          findVal(cityData);
        }
      });
    }
  }
};

exports.default = SelectCity;
module.exports = exports['default'];
